### PR TITLE
ci: refactor 'Main' github workflow

### LINKED ISSUE #3999: Improve the runtime for the CI check 'Main'
**Describe the feature request**
Successful completion of 'Main' github workflow ranges from ~ 1h 30m to ~1h 50m.
We run two parallel jobs in this workflow:
image

In build step, following steps are linear and takes time:
image

These steps can be run in parallel.
The lint step is running in parallel anyway.

Proposal:
Run all the 4 operations into different github workflows altogether for better reporting and signals.

**Describe alternative(s) you've considered**

**Affected product area (please put an X in all that apply)**

- ( ) APIs
- ( ) Addons
- ( ) CLI
- ( ) Docs
- ( ) IAM
- ( ) Installation
- ( ) Plugin
- ( ) Security
- ( ) Test and Release
- ( ) User Experience
- (x) Developer Experience

**Additional context**

### PR DESCRIPTION
**Pull request by navidshaikh**
### What this PR does / why we need it
Improving the Main CI operation runtime by ~30%

    ci: refactor 'Main' github workflow

    Earlier main.yaml workflow was running 2 jobs; build and lint.
    build job was performing: build, verify and test steps in
    sequence, while lint job was running in parallel with build.

    This changeset splits main.yaml workflow into 4 separate workflows;
    build, test, verify and lint for running them in parallel.
    Also updates the Makefile to run 'make tools' target before 'make verify'.

    Signed-off-by: Navid Shaikh <navids@vmware.com>

### Which issue(s) this PR fixes

Fixes #3999 

### Describe testing done for PR
Check test PR https://github.com/vmware-tanzu/tanzu-framework/pull/3996 

all 4 workflows running in parallel and reduced time to 59 mins, which is equal to the workflow running longest (test)
image

or check the checks in this PR https://github.com/vmware-tanzu/tanzu-framework/pull/4000#issuecomment-1330110844

### Release note

```release-note
CI: Refactor github workflow 'Main' to improve the workflow execution time
```

### CODE REVIEW

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -0,0 +1,94 @@
+name: Build
+
+on:
+  pull_request:
+    branches: [ pr/optimize-workflows ]
```
Comment (navidshaikh) [inline .github/workflows/build.yaml:5]: need to update the branch to main and release-* once the review is complete (using this branch to test the workflows)

File: .github/workflows/verify.yaml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      run: |
+        echo "::set-output name=dir::$(npm config get cache)"
+    - name: npm cache
+      uses: actions/cache@v3
+      id: npm-cache # use this to check for `cache-hit` ==> if: steps.npm-cache.outputs.cache-hit != 'true'
+      with:
+        path: ${{ steps.npm-cache-dir.outputs.dir }}
+        key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
+        restore-keys: |
+          ${{ runner.os }}-node-
```
Comment (navidshaikh) [inline .github/workflows/verify.yaml:55]: think this part can be removed since we're not doing any node related ops in make verify

Comment (navidshaikh) [reply .github/workflows/verify.yaml]: updated in 
2f174eb

**Comment by rajathagasthya** [comment]
Can we make all these checks mandatory to merge a PR (if it's not done already)?

**Comment by navidshaikh** [comment]
Thanks for the review @rajathagasthya! 

> Can we make all these checks mandatory to merge a PR (if it's not done already)?

Sure, note that currently the Main workflow checks aren't configured mandatory in branch protection rules but only 'Check labels' check, will mark as mandatory once the PR is in

**Comment by vuil** [comment]
Thanks @navidshaikh changes makes sense, now that even make verify is taking upwards of 15m. 
Why are the updated checks not being run in this PR though?

**Comment by navidshaikh** [comment]
> I think the workflows yaml files should be committed in the base branch for running (se https://github.com/vmware-tanzu/tanzu-framework/pull/3996)

scratch that, configured 'main' (and release-*) branches in the PR branch and they're running @vuil

**Comment by navidshaikh** [comment]
checks passing 
image
 will rebase/squash to get this ready for merge

File: .github/workflows/build.yaml (hunk 1 above)
Comment (navidshaikh) [reply .github/workflows/build.yaml]: updated, ready for review/merge

**Review by randomvariable — APPROVED** [review]

**Comment by navidshaikh** [comment]
marked build, lint, unittests and verify checks as mandatory for main and release branches cc: @rajathagasthya

**Comment by navidshaikh** [comment]
Reverted the change to mark the checks as required since the outstanding PRs, for which CI didnt run were blocked for un-triggered required checks, causing confusion. New PRs or rebased ones will get the new checks, older PRs where rebase is costly can be merged. We'll mark the new checks as required later.

**Pull request merged**